File: Python/Insertionsort.py
"""
Insertionsort
We iterate over all elements and place them in the correct new order.
We create the gaps for the insertions by shifting the elements one to the right.
Worst case runtime O(n^2)
"""


def insertion_sort(arr):
    n = len(arr)
    for i in range(1, n):
        elem = arr[i]  # remember our current element
        j = i  # we guess that our element is already in the correct place
        while j > 0 and arr[j - 1] > elem:  # Move greater elements than our current value one position ahead
            arr[j] = arr[j - 1]
            j -= 1
        arr[j] = elem
    return arr


if __name__ == "__main__":
    test_arr = [
        3,
        1,
        7,
        4,
        12,
        9,
        11,
        102
    ]

    print(insertion_sort(test_arr))
    import benchmark
    benchmark.benchmark(insertion_sort)

File: C++/Insertionsort.cpp
#include <iostream>
#include <vector>

#include "Randomdata.cpp"

void insertionSort(std::vector<int> &arr){
    int n = arr.size();
    for (int i = 1; i < n; i++){
        int elem = arr[i];
        int j = i;
        while (j > 0 && arr[j - 1] > elem){
            arr[j] = arr[j - 1];
            j--;
        }
        arr[j] = elem;
    }
}

void printArray(const std::vector<int>& arr){
    for (auto const& v: arr){
        std::cout << v << " | ";
    }
}

int main(){
    std::vector<int> arr = {12, 321, 12, 1, 5, 34, 27};
    insertionSort(arr);
    printArray(arr);

    std::vector<int> bench = randomData(5000);
    insertionSort(bench);
    printArray(bench);

    return 0;
}